src/llfs/ioring_log_device2.hpp

```diff
@@ -0,0 +1,498 @@
+#ifndef LLFS_IORING_LOG_DEVICE2_HPP
+#define LLFS_IORING_LOG_DEVICE2_HPP
+
+#include <llfs/config.hpp>
+//
+#include <llfs/basic_ring_buffer_log_device.hpp>
+#include <llfs/ioring_log_config2.hpp>
+#include <llfs/ioring_log_device_storage.hpp>
+#include <llfs/log_device.hpp>
+#include <llfs/log_device_runtime_options.hpp>
+#include <llfs/packed_log_control_block2.hpp>
+#include <llfs/ring_buffer.hpp>
+
+#include <batteries/async/watch.hpp>
+#include <batteries/math.hpp>
+#include <batteries/status.hpp>
+#include <batteries/tuples.hpp>
+
+#include <chrono>
+#include <deque>
+#include <thread>
+#include <vector>
+
+namespace llfs {
+
+BATT_STRONG_TYPEDEF(slot_offset_type, TargetTrimPos);
+BATT_STRONG_TYPEDEF(slot_offset_type, CommitPos);
+
+/** \brief Initializes an IoRingLogDevice2 using the given storage device and configuration (which
+ * includes offset within the passed device).
+ */
+Status initialize_log_device2(RawBlockFile& file, const IoRingLogConfig2& config);
+
+template <typename StorageT>
+class IoRingLogDriver2
+{
+ public:
+  using Self = IoRingLogDriver2;
+  using AlignedUnit = std::aligned_storage_t<kLogAtomicWriteSize, kLogAtomicWriteSize>;
+  using EventLoopTask = typename StorageT::EventLoopTask;
+
+  static constexpr batt::StaticType<TargetTrimPos> kTargetTrimPos{};
+  static constexpr batt::StaticType<CommitPos> kCommitPos{};
+
+  //+++++++++++-+-+--+----- --- -- -  -  -   -
+
+  /** \brief The size (bytes) of each preallocated completion handler memory object.
+   */
+  static constexpr usize kHandlerMemorySize = 160;
+
+  using HandlerMemory = batt::HandlerMemory<kHandlerMemorySize>;
+
+  using HandlerMemoryStorage =
+      std::aligned_storage_t<sizeof(HandlerMemory), alignof(HandlerMemory)>;
+
+  //+++++++++++-+-+--+----- --- -- -  -  -   -
+
+  explicit IoRingLogDriver2(LogStorageDriverContext& context,        //
+                            const IoRingLogConfig2& config,          //
+                            const LogDeviceRuntimeOptions& options,  //
+                            StorageT&& storage                       //
+                            ) noexcept;
+
+  //----
+
+  Status set_trim_pos(slot_offset_type trim_pos)
+  {
+    this->observed_watch_[kTargetTrimPos].set_value(trim_pos);
+    return OkStatus();
+  }
+
+  slot_offset_type get_trim_pos() const
+  {
+    return this->trim_pos_.get_value();
+  }
+
+  StatusOr<slot_offset_type> await_trim_pos(slot_offset_type min_offset)
+  {
+    return await_slot_offset(min_offset, this->trim_pos_);
+  }
+
+  //----
+
+  slot_offset_type get_flush_pos() const
+  {
+    return this->flush_pos_.get_value();
+  }
+
+  StatusOr<slot_offset_type> await_flush_pos(slot_offset_type min_offset)
+  {
+    return await_slot_offset(min_offset, this->flush_pos_);
+  }
+
+  //----
+
+  Status set_commit_pos(slot_offset_type commit_pos)
+  {
+    this->observed_watch_[kCommitPos].set_value(commit_pos);
+    return OkStatus();
+  }
+
+  slot_offset_type get_commit_pos() const
```

Comment:
The `set_`...  methods for log offset pointers can fail because the new value might violate an invariant (e.g., trim moves past flush or commit, resulting in a "negative" size, or commit pos is set greater than trim pos plus log capacity), and also because the log device can be in a "closed" or inactive state.  Getters don't have the first kind of problem (since they by definition are read-only), and by convention the current implementation allows reading the last valid value even after closing/halting the log device.